sortingHat/sortingHat.py

```diff
@@ -0,0 +1,347 @@
+#May be more applicable to a hungarian algorithm approad
+def calcStudentPreference(known, learn, preference):
+    # weight variables. This should be more extensible to account
+    # for multiple possible variables
+    assert (known >= 0)
+    assert (learn >= 0)
+    assert (preference > 0)
+    pref = 5 * 1/preference + LEARN_WEIGHT * learn + \
+            KNOWN_WEIGHT * known
+    pref = pref/(1 + LEARN_WEIGHT + KNOWN_WEIGHT)
+    # return a value normalized between 0 and 1
+    return (pref - 1)/(8 - 1)
+
+##! Use this to make list a group preferences for gale-shapley
+def calcGroupPreference(known, learn):
+    assert (known >= 0)
+    assert (learn >= 0)
+    pref = LEARN_WEIGHT * learn + KNOWN_WEIGHT * known
+    pref = pref/(KNOWN_WEIGHT + LEARN_WEIGHT)
+    return pref/3 # group pref range is 0-3, so normalize between 0 and 1
+
+##! This function registers students with each connected group and adds the
+##! student to the group's preference list with the associated preference score
+def registerUser(student):
+    known_score = 0
+    learn_score = 0
+    for grp in student.preferences:
+        # How to weight group preferences for students?
+        for attr in grp.skills_needed:
+            if attr in student.known_skills and known_score < 10:
+                known_score += 1 
+            if attr in student.skills_to_learn and learn_score < 10:
+                learn_score += 1
+        grp.preferences[student] = calcGroupPreference(known_score, learn_score)
+        
+
+def trimGroups():
+    g_count = int(float(len(students))/OPT_SIZE)
+    min_edges = 10000 
+    while len(groups) > g_count:
+        for group in groups:
+            if len(group.preferences) < min_edges:
+                group_to_drop = group
+        print("Removing %s" % group)
+        groups.remove(group)
+    for student in students:
+        if group in student.preferences:
+            student.preferences.remove(group)
+
+def unpopular():
+    for group in groups:
+        if len(group.preferences) < MIN_SIZE:
+            print(group.name + " is unpopular")
+
+def swapThemBitches(shortGroup, firstPass):
+    print("Fixing " + shortGroup.name)
+    for group in groups:
+        if group == shortGroup:
+            continue
+        if len(group.members) > OPT_SIZE:
+            for student in group.members:
+                if len(shortGroup.members) >= MIN_SIZE:
+                    return
+                if shortGroup in student.preferences:
+                    group.members.remove(student)
+                    shortGroup.members.append(student)
+        elif not firstPass and len(group.members) > MIN_SIZE:
+            for student in group.members:
+                if len(shortGroup.members) >= MIN_SIZE:
+                       return
+                if shortGroup in student.preferences:
+                       group.members.remove(student)
+                       shortGroup.members.append(student)
+
```

Comment:
Correct me if I'm wrong, but as I'm understanding this we've only got 2 chances to make even sized groups? And if after !firstPass the groups aren't the same size then oh well?

Also, unrelated, but I think we should add to our ice box tasks to break the SortThemBitches function up into sub-functions so it's easier to parse (and possibly reusable!). I don't think this is necessary for the first commit though, and I think you could delegate this to someone else (like another algorithm team member, or me ;) )